keyboards/crkbd/keymaps/emdashii/oled.c

```diff
@@ -0,0 +1,143 @@
+/* Copyright 2022 @toinux
+  * Copyright 2022 @emdashii
+  * This program is free software: you can redistribute it and/or modify
+  * it under the terms of the GNU General Public License as published by
+  * the Free Software Foundation, either version 2 of the License, or
+  * (at your option) any later version.
+  *
+  * This program is distributed in the hope that it will be useful,
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  * GNU General Public License for more details.
+  *
+  * You should have received a copy of the GNU General Public License
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
+
+
+#include QMK_KEYBOARD_H
+#include "keycodes.h"
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+  if (!is_keyboard_master()) {
+    return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
+  }
+  return rotation;
+}
+
+// #define _QWERTY 0
+// #define _NAVIGATION 1
+// #define _GAMING 2
+// #define _NUMBER 3
+// #define _SYMBOL 4
+// #define _FUNCTION 5
+// #define _ADJUST 6
+
+// This renders, but the switching function is very broken.
+void oled_render_layer_state(void) {
+    oled_write_P(PSTR("\nLayer: "), false);
+    switch (get_highest_layer(layer_state|default_layer_state)) {
+        case _QWERTY:
+            oled_write_ln_P(PSTR("QWERTY"), false);
+            break;
+        case _NAVIGATION:
+            oled_write_ln_P(PSTR("Navigation"), false);
+            break;
+        case _GAMING:
+            oled_write_ln_P(PSTR("Gaming"), false);
+            break;
+        case _NUMBER:
+            oled_write_ln_P(PSTR("Numbers"), false);
+            break;
+        case _SYMBOL:
+            oled_write_ln_P(PSTR("Symbols"), false);
+            break;
+        case _FUNCTION:
+            oled_write_ln_P(PSTR("Fn Keys"), false);
+            break;
+        case _ADJUST:
+            oled_write_ln_P(PSTR("Settings"), false);
+            break;
+        default:
+            oled_write_ln_P(PSTR("Default"), false);
+            break;
+    }
+}
+
+
+//
+char keylog_str[24] = {};
+
+const char code_to_name[60] = {
+    ' ', ' ', ' ', ' ', 'a', 'b', 'c', 'd', 'e', 'f',
+    'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p',
+    'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+    '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+    'R', 'E', 'B', 'T', '_', '-', '=', '[', ']', '\\',
+    '#', ';', '\'', '`', ',', '.', '/', ' ', ' ', ' '};
+
+uint8_t last_row = 0;
+uint8_t last_col = 0;
+
+void set_keylog(uint16_t keycode, keyrecord_t *record) {
+  char name = ' ';
+    if ((keycode >= QK_MOD_TAP && keycode <= QK_MOD_TAP_MAX) ||
+        (keycode >= QK_LAYER_TAP && keycode <= QK_LAYER_TAP_MAX)) { keycode = keycode & 0xFF; }
+  if (keycode < 60) {
+    name = code_to_name[keycode];
+  }
+
+  // update keylog
+  snprintf(keylog_str, sizeof(keylog_str), "%dx%d, k%2d : %c",
```

Comment:
https://docs.qmk.fm/#/squeezing_avr?id=oled-tweaks

This uses ~1.5kb less firmware space.

Reply:
Correct. the snprintf function takes up a significant amount of speace. 

As for a keymap that has already done this, not for the corne specifically.  At least, not that I can remember.  But the above should outline that enough. 

```c
  snprintf(keylog_str, sizeof(keylog_str), "%dx%d, k%2d : %c",
           record->event.key.row, record->event.key.col,
           keycode, name);
```
would become something like: 
```c
    oled_write(get_u8_str(record->event.key.row, ' '), false);
    oled_write_P(PSTR("x"), false);
    oled_write(get_u8_str(record->event.key.col, ' '), false);
    oled_write_P(PSTR(", k"), false);
    oled_write(get_u16_str(keycode, ' '), false);
    oled_write_P(PSTR(" : "), false);
    oled_write(name, false);
```

If you want shorting strings (get_u8_str uses 3 characters, get_u16_str uses 5), you can implement a custom implementation.  You can see how the functions are assembled here: 
https://github.com/qmk/qmk_firmware/blob/40d295abf69a79f67628e06599bfdbe50169aa4c/quantum/quantum.c#L519-L573

Reply:
Thanks for the detailed response, I got the code working. I don't know what the best practice is to add this change to the pull request on github. I think I added it, but I'm not sure.

Reply:
Welcome!  And it isn't best practice, specifically, but on any AVR processor, very byte you can get back is worth it, especially on a board like the Corne.